---
title: "Get Epigenome RoadMap Data"
author: "Rick Farouni"
date: "November 17, 2017"
output:
  html_document:
    df_print: paged
    code_folding: show
    toc: yes
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

### Install Bioconductor Packages


```{r, eval=FALSE}
#source("https://bioconductor.org/biocLite.R")
#biocLite(c("biomaRt",
#           "AnnotationHub",
#           "BSgenome.Hsapiens.UCSC.hg19",
#           "ChIPQC",
#           "GenomicFiles"))

```

### Load Packages

```{r}
library(AnnotationHub)
library(biomaRt)
```

```{r, message=FALSE, warning=FALSE}
library(dplyr)
library(tibble)
library(tidyr)
library(readr)
#library(data.table)
```


### Set directory paths

```{r}
#proj_dir <- "/mnt/hd2/Dropbox (Partners HealthCare)/projects/2017_10_histone_code/RMEJ_analysis"
root_dir <- "/mnt/hd2/projects/histone_code/histone_code"
annotations_dir <- file.path(root_dir,
                             "data/annotations")
proj_dir <- file.path(root_dir, "preprocessing")
setwd(proj_dir)
sample_names_dir <-  file.path(proj_dir,  "sample_names")
#dir.create(sample_names_dir)
output_dir <- file.path(proj_dir,"files_source_urls")

data_dir <- "/data/molpath/common_data/roadmap"
bigwig_dir <- file.path(data_dir, "bigwig_pvalue") 
rpkm_output_dir <- file.path(data_dir, "gene_expression" ,"rpkm")


#dir.create(rpkm_output_dir)
```

### Download Gene expression metadata


```{r, eval=FALSE}

download.file("http://egg2.wustl.edu/roadmap/data/byDataType/rna/expression/EG.name.txt",
              "EG.name.txt")
```


### Read data

```{r}
eg_names <- data.table::fread(input = "EG.name.txt", 
                    header = FALSE, 
                    col.names = c("epigenome", 
                                  "celltypes")) %>% 
  filter(epigenome != "E000" )

eg_names 
```


## Get experiments metada from RoadMap Epigenomics

```{r , message=FALSE, warning=FALSE}

ah <- AnnotationHub()

# Searching for  epigenomes in the roadmap database

epiFiles <- query(ah, c("EpigenomeRoadMap", "^E"))

epiFiles <- query(epiFiles , c( "H3K27ac.pval.signal.bigwig",
                                "H3K27me3.pval.signal.bigwig",
                                "H3K4me3.pval.signal.bigwig",
                                "H3K36me3.pval.signal.bigwig",
                                "H3K9me3.pval.signal.bigwig",
                                "H3K4me1.pval.signal.bigwig"),
                  pattern.op = `|`)

epiFiles

```


## Create  epigenome-by-mark array 

```{r}
RMEP_dt <- data.frame(title = epiFiles$title,
                      sourceurl = epiFiles$sourceurl) %>%
    bind_cols(do(., data.frame(
      str_split(.$title,
                "-|.pval",
                simplify = TRUE)[,1:2],
      stringsAsFactors = FALSE)))  %>%
  select(-title) %>%
  rename( epigenome= "X1",
          mark= "X2") 

marks <- RMEP_dt %>% 
  select(mark) %>% 
  unique  %>% 
  unlist

RMEP_dt
```



```{r}
epifiles_dt <- RMEP_dt %>% 
  data.table::dcast( epigenome ~ mark )

epifiles_dt
```

## Filter Epigenomes

Epigenomes that have gene expression data

```{r , message=FALSE, warning=FALSE}
epifiles_dt  <- left_join(eg_names,
                          epifiles_dt,
                      by ='epigenome')

epifiles_dt
```


Number of epigenomes with gene expression data that have 1, 2, 3, or 4 of the marks (DNase, H3K27ac, H3K27me3 ) 
```{r}
epifiles_dt$marks_num <- epifiles_dt %>%
     select(-epigenome, -celltypes)  %>%
     is.na %>% 
    `!` %>% 
    rowSums 

epifiles_dt
```

### Frequency counts of marks

```{r}
epifiles_dt %>% 
  select(marks_num) %>% 
  table

```


Epigenomes with all 6 marks



```{r}
selected_epigenomes <- epifiles_dt %>%
  filter(marks_num == 6) %>%
  filter(!(epigenome %in% c("E056", "E059"))) %>%
  arrange(marks_num) 

selected_epigenomes$epigenome_name <- lapply(
  str_split(
    selected_epigenomes$celltypes,
    "_Derived|_Cultured_Cells|_Cell_Line|_Primary_Cells|_Foreskin|_Naive_Primary_Cells|_skin02|_skin03|_skin01|_Foreskin|_Cells|_Female|_Mononuclear|_Middle",
    n = 4),
  str_c,
  collapse = "") %>% 
  unlist %>% 
  str_c(selected_epigenomes$epigenome,
                   .,
                   sep = "_")


selected_epigenomes 
```


###  Save sample name metadata to file if RPKM exits


```{r}

sample_names_dt <- list()

for (mark_target in marks) {
  sample_names_dt[[mark_target]] <- selected_epigenomes  %>%
    select(epigenome_name,
           mark_target,
           epigenome)  %>%
    rowwise() %>%
    mutate(
      mark = mark_target,
      epigenome_mark = str_c(epigenome, "_", mark_target),
      filepath =  file.path(
        bigwig_dir,
        paste0(epigenome, "-", mark_target, ".pval.signal.bigwig")
      ),
      rpkm_filepath = file.path(rpkm_output_dir,
                                paste0(epigenome,
                                       "_avg_rpkm.txt"))
    ) %>%
    select(epigenome_mark,
           epigenome_name,
           epigenome,
           mark,
           filepath,
           rpkm_filepath)
}

metadata_dt <- do.call(rbind, sample_names_dt)

metadata_dt
```


## save to file

```{r, eval=FALSE}
data.table::fwrite(
  metadata_dt,
  file.path(annotations_dir,"rmep_metadata_46epigenomes_6histones.txt"),
  col.names = FALSE,
  sep = "\t")
```

